### ISSUE TITLE
Any way to add node to center of canvas?

### ISSUE THREAD

**Issue by gappleto97** [anvaka/VivaGraphJS#153]
Sorry if I'm missing something obvious, but as near as I can tell, the point of insertion for nodes is wandering all over the place. I just want a reliable way to add nodes to the center of any given canvas, and have them rearrange accordingly afterwards.

Is there a way to do this?

**Comment by anvaka** [comment]
I believe this is a miss on my end - I cannot find this documented in the current repository.

What you are looking for is `layout.setNodePosition(nodeId, x, y)` - the method will override default node position and set it to the given `x` and `y`.

It was documented in the [ngraph.forcelayout](https://github.com/anvaka/ngraph.forcelayout#pin-node-and-initial-position) module, but not here.

Please let me know if this works for you.

**Comment by gg4u** [comment]
I have the opposite problem: I want to add neighbors using as relative center the parent node, not the center of the canvas.
in my test graph.addLinks() works with this behavior but if I use them within an AJAX call, the new nodes will appear on the center of the canvas, and will be drag around afterwards with no nice effect.... 

While I am writing this, I thought that maybe the solution could be move graph.addLinks() out of success function in AJAX, and put it within a callback. (If fetching neighbors is successfull, then place them ....).

However, I would like to know if you have experience this behavior, too:
have you experienced a different positioning of nodes when dynamically adding neighbors within AJAX call?

**Comment by gappleto97** [comment]
I'm not exactly sure how to call this method. When I try to use the ngraph syntax, it says that "require is not defined". When I try to access it via "graph.layout", it's undefined.

Can you cure my stupid?

**Comment by gappleto97** [comment]
Just for reference, this is me trying to solve [Issue #2](https://github.com/gappleto97/visualizer/issues/2) on my repo.

In addition to my comment above, I've tried going into each node ad changing their position to {x:0,y:0}, but it appears that this location still drifts over time.

**Comment by gappleto97** [comment]
using the method you described, it's still drifting. I'm currently calling layout.setNodePosition(nodeId,0,0); on every addition, yet this point still drifts significantly over time.

I'm going to try a workaround to see if I can give you more data to work off of. I'll get back to you in a bit.

**Comment by gappleto97** [comment]
Just found out that layout.setNodePosition() was calling a non-existent method, so that would explain some. I decided to try going in and replacing force_directed_bodies properties, and that works for a little while, but the program quickly crashes. Still, though, there's no noticeable drift in that time.

**Comment by gappleto97** [comment]
Wow, so I'm a little embarrassed here.

Turns out when I forked this project, I never checked what version of Vivagraph it was using. It was nearly 2 years old. So yeah, my fault, not yours. I'm still having trouble properly placing 2/4 of my node types, but that's on me, not you.

I still wish there was a way to automatically do this, but feel free to consider this closed.

**Comment by anvaka** [comment]
Hi @gappleto97 

The layout object created [here](https://github.com/anvaka/VivaGraphJS/blob/ca953396a8f0ed5a007f0f47188b7c5abfa6723b/demos/other/physicsLayout.html#L20-L25) should be exactly the same as `ngraph.forcelayout` ([see here](https://github.com/anvaka/VivaGraphJS/blob/cab1fca26d74713f8b556891f615e08b766c60cb/src/viva.js#L64) how it's created internally). So documentation for `ngraph.forcelayout` applies to vivagraph too.

> Still, though, there's no noticeable drift in that time.

I'm not sure I understand. Are you trying to prevent nodes from flying away?

**Comment by gappleto97** [comment]
No. When I call addNode(), the location it adds them at changes over time.
I'm trying to prevent this from happening, and add to the center of the
canvas.

Using the above hack makes it relatively stable, but for some reason I'm
not catching all of the nodes I'm adding. And even if I was, it's still
wasting processor time to do it. Center by default would be more efficient.
Is there some obvious way that I'm missing? Could I perhaps pass in a
force_directed_body object on creation?

**Comment by anvaka** [comment]
Ah, I see what you mean. 

Physics simulator will attempt to to put node as close as possible towards its neighbors. This is by design, since in most cases it will reduce amount of required work. 

What this means - if you add your node via `addLink()` it will use connection information to position those nodes.

Have you tried calling `layout.setNodePosition(nodeId, 0, 0)`, to override default initial position with the latest version of the library? What happens?

**Comment by gappleto97** [comment]
I have. I haven't tried it on the most recent version yet (again, was using
a 2 year old one...), but in that one the location would also shift over
time. It was quite confusing.

I'll try it on the new version tomorrow. Hopefully that will work better.
